Factor graph out-of-core learning and inference

igozali opened the pull request:
Provides support for learning and inference on factor graphs that don't fit inside memory. Changes in DeepDive are required to support this feature.

igozali committed:
Revert "Drops incomplete checkpoint/resume code to first merge the refactor"

Restore checkpoint/resume code to work on it.

Signed-off-by: Ivan Gozali <igozali@stanford.edu>

igozali committed:
Disable checkpoint_test until checkpoint/resume is somewhat implemented.

Signed-off-by: Ivan Gozali <igozali@stanford.edu>

igozali committed:
Fixed checkpoint/resume code for CompactFactorGraph.

It's true that the refactored code caused this checkpoint/resume code
to make much more sense.

Signed-off-by: Ivan Gozali <igozali@stanford.edu>

igozali committed:
Checkpoint/resume very nicely fits in the new code.

Signed-off-by: Ivan Gozali <igozali@stanford.edu>

igozali committed:
Fixes RAII issues and clean up CFG resumption.

Signed-off-by: Ivan Gozali <igozali@stanford.edu>

igozali committed:
Do not reimplement mkdir -p

Signed-off-by: Ivan Gozali <igozali@stanford.edu>

igozali committed:
Don't waste up space in classes for string literals.

My bad, I must have been in Java mode when I wrote that.

Signed-off-by: Ivan Gozali <igozali@stanford.edu>

igozali committed:
Unused function declarations

Signed-off-by: Ivan Gozali <igozali@stanford.edu>

igozali committed:
Tried to rip off Weights from *FactorGraph

Signed-off-by: Ivan Gozali <igozali@stanford.edu>

igozali commented:
Note that this feature is far from done; @netj and I thought it would be easier for everyone to chip in to give comments as well as the code is starting to materialize.

igozali committed:
Travis doesn't like auto loop variables.

Signed-off-by: Ivan Gozali <igozali@stanford.edu>

igozali committed:
Clarify pointer vs. array

Signed-off-by: Ivan Gozali <igozali@stanford.edu>

igozali commented on src/factor_graph.h:
```diff
@@ -33,7 +33,10 @@ class FactorGraphDescriptor {
+   * number of weights. this is still kept here to maintain format of meta,
+   * and is also useful when loading weights
+   */
   num_weights_t num_weights;
```
Here's another FactorGraph dependency with Weights. When calling InferenceResult's constructor, we pass in the FactorGraph and Weights array. InferenceResult preallocates the weights_{isfixed,value} arrays based on the FactorGraphDescriptor inside FactorGraph.

Again, this could be solved with creating a new LearningState class, I suppose? For now, I'll make things work with two floating information (num_weights and weights), and then later group num_weights and weights into a LearningState object. 

@netj I'm getting the sense that you don't like the std::vector idea?

netj replied:
Ideally, these shall get decoupled and InferenceResult (or LearningState) will be constructed independently from others (e.g., FactorGraph or GibbsSampler).

Not sure how much overhead vector has over plain array access, but I think we shouldn't give up anything since weight access is on a really hot path.

netj commented on src/inference_result.cc:
```diff
@@ -142,7 +144,8 @@ void InferenceResult::dump_weights_in_binary(std::ofstream &binary_output) const
 // FIXME: See note on load_assignments.
-void InferenceResult::dump_assignments_in_binary(std::ofstream &binary_output) const {
+void InferenceResult::dump_assignments_in_binary(
+    std::ofstream &binary_output) const {
```
In general I think `ostream` is more reusable than `ofstream`.  Not sure what you can do more with the `f`.

igozali committed:
Make tests work again.

Turns out this is nontrivial, and shows how the interface is somewhat
broken right now if we naively just rip off Weights from FG.

Signed-off-by: Ivan Gozali <igozali@stanford.edu>
